hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpConnector.java

```diff
@@ -163,43 +163,50 @@ private void tryToConnect(InetSocketAddress socketAddress, int timeout) throws E
 
-            try {
-                channel.connect(socketAddress, timeout);
-
-                ioService.interceptSocket(socketChannel.socket(), false);
-
-                TcpIpConnection connection = connectionManager.newConnection(channel, address);
-                connectionManager.sendBindRequest(connection, address, true);
-            } catch (NullPointerException e) {
-                // Helper piece of code, which will allow to identify rare NPEs in TLS connections
-                // https://github.com/hazelcast/hazelcast-enterprise/issues/2104
-                //TODO remove this catch block once the TLS NPE problem is successfully resolved
-                closeSocket(socketChannel);
-                logger.log(level, "Could not connect to: " + socketAddress + ". Reason: " + e.getClass().getSimpleName()
-                        + "[" + e.getMessage() + "]");
-                logger.log(Level.INFO,
-                        "Add this stacktrace to https://github.com/hazelcast/hazelcast-enterprise/issues/2104 please!", e);
-                throw e;
-            } catch (Exception e) {
-                closeSocket(socketChannel);
-                logger.log(level, "Could not connect to: " + socketAddress + ". Reason: " + e.getClass().getSimpleName()
-                        + "[" + e.getMessage() + "]");
-                throw e;
+                try {
+                    channel.connect(socketAddress, timeout);
+
+                    ioService.interceptSocket(socketChannel.socket(), false);
+
+                    TcpIpConnection connection = connectionManager.newConnection(channel, address);
+                    connectionManager.sendBindRequest(connection, address, true);
+                    success = true;
+                } catch (NullPointerException e) {
+                    // Helper piece of code, which will allow to identify rare NPEs in TLS connections
+                    // https://github.com/hazelcast/hazelcast-enterprise/issues/2104
+                    //TODO remove this catch block once the TLS NPE problem is successfully resolved
+                    closeSocket(socketChannel);
+                    logger.log(level, "Could not connect to: " + socketAddress + ". Reason: " + e.getClass().getSimpleName()
+                            + "[" + e.getMessage() + "]");
+                    logger.log(Level.INFO,
+                            "Add this stacktrace to https://github.com/hazelcast/hazelcast-enterprise/issues/2104 please!", e);
+                    throw e;
+                } catch (Exception e) {
+                    closeSocket(socketChannel);
+                    logger.log(level, "Could not connect to: " + socketAddress + ". Reason: " + e.getClass().getSimpleName()
+                            + "[" + e.getMessage() + "]");
+                    throw e;
+                }
+            } finally {
+                if (!success) {
+                    connectionManager.acceptedChannels.remove(channel);
```

Comment:
Could you wrap this into a method? It would be a cleaner way. 

I think, the `!success` check is not needed here - the remove will take care of it (i.e. we want to remove it in either case). If the `connectionManager.newConnection()` implementation is changed in a future and it doesn't remove the channel, this would be a guard.

Reply:
Done